staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go

```diff
@@ -104,11 +105,19 @@ func (s *ServerRunOptions) Validate() []error {
 	if s.TargetRAMMB < 0 {
 		errors = append(errors, fmt.Errorf("--target-ram-mb can not be negative value"))
 	}
-	if s.MaxRequestsInFlight < 0 {
-		errors = append(errors, fmt.Errorf("--max-requests-inflight can not be negative value"))
-	}
-	if s.MaxMutatingRequestsInFlight < 0 {
-		errors = append(errors, fmt.Errorf("--max-mutating-requests-inflight can not be negative value"))
+
+	if s.EnableInfightQuotaHandler {
+		if s.MaxMutatingRequestsInFlight != 0 || s.MaxRequestsInFlight != 0 {
```

Comment:
Won't the default values make this fail?

Reply:
yeah.. it can indeed surprise the users a bit when they try to enable the new handler. i updated the error msg to print the current value of the max-inflight options so that it helps users to get know that the legacy flag is set by default. and in the future, if the new handler goes stable, wydt if we change the default the new handler to enabled and meanwhile default `--max-mutating-requests-inflight` and `--max-requests-inflight` to zero? not sure it's losing backward-compatibility to adjust the default value for old falgs?

Reply:
Let's not forget that our new handler also needs to know the total server concurrency to divide up among the RequestPriorities.  And we need an evolutionary path to that.

Following is my suggested path.
Step 1: make no changes to the max-in-flight handler's flags (not even their default values); the new handler uses as its total server concurrency limit the sum of the old handler's two concurrency limits.

Step 2: introduce a new flag, with default 0, that MAY be used to specify the total server concurrency limit for our new handler.  If the new flag is not supplied then the new handler continues to use the sum of the old handler's limits.  The old handler is modified so that if it is enabled and the new flag is given and the old flags have zero values, the new handler's limit is divided up to produce the limits that the old handler will use --- 2/3 for non-mutating and 1/3 for mutating.

Step 3: switch the defaults, so that the old flags default to zero and the new flag defaults to 600.

Step 4: de-implement the old flags.

Reply:
>  The old handler is modified so that if it is enabled and the new flag is given and the old flags have zero values, the new handler's limit is divided up to produce the limits that the old handler will use --- 2/3 for non-mutating and 1/3 for mutating. 

it doesn't seem necessary to me.. why does the old handler respect the new flag? 

> Make validation of flags check that if both old and new flags are given then the new flag is the sum of the old two.

for a simplicity, i presume it's better if the new flag only works for the new handler and remain the old one the same. summing them up make it a bit complex.

fwiw, what do you think if we just start by introducing only one new feature gate and one new flag for switching handler in this pull? as for another new flag specifying concurrency limit, let's land that onto feature branch?

Reply:
What I outlined is an evolutionary progression in how the server's concurrency limit(s) are stated such that at every step along the way, we get good results even if the command line parameters for concurrency limit(s) are specified according to the previous step or the one before that.  This is to support backwards compatability.  I am thinking that the two handlers --- max-in-flight and our new one --- are two different ways of managing against the server's concurrency limit(s), and the choice of management technique is an independent issue from the specifying of the limit(s).  I am trying to minimize disruption to how clusters are configured.  Remember that it is not always done by humans, it is more and more often something that is automated --- and automation is often dumber, more inflexible, and less up-to-date than humans.

Reply:
And yes, I think it would be fine to put the new concurrency limit flag in the feature branch.